module/device/connection.py

```diff
@@ -290,8 +291,14 @@ def _nc_server_host_port(self):
         # For emulators, listen on current host
         if self.serial.startswith('emulator-') or self.serial.startswith('127.0.0.1:'):
             host = socket.gethostbyname(socket.gethostname())
+            if platform.system() == 'Linux' and host == '127.0.1.1':
+                host = '127.0.0.1'
             logger.info(f'Connecting to local emulator, using host {host}')
             port = random_port(self.config.FORWARD_PORT_RANGE)
+            board = self.adb_shell(['getprop', 'ro.hardware.audio.primary'])
+            hardware = self.adb_shell(['getprop', 'ro.hardware'])
+            if 'goldfish' in board or "ranchu" in hardware:
+                return host, port, "10.0.2.2", port
```

Comment:
This will cost 30ms, and will cost more on slow emulators. Reduce adb_shell calls like
```python
if serial_may_avd(self.serial):
    board = self.adb_shell(['getprop', 'ro.hardware.audio.primary'])
    if 'goldfish' in board:
        return xxx
    hardware = self.adb_shell(['getprop', 'ro.hardware'])
    if "ranchu" in hardware:
        return xxx
return host, port, host, port
```

Reply:
AVD uses `127.0.0.1:5555` or `emulator-5554` (so exactly the same as LDPlayer)